### PR TITLE
syntax: Make imports in AST closer to the source and cleanup their parsing

### PR DESCRIPTION
**Pull request by petrochenkov**
This is a continuation of https://github.com/rust-lang/rust/pull/45846 in some sense.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Could someone else from @rust-lang/compiler take this over? This touches a lot of code that I know little about.

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -1880,18 +1880,29 @@ pub type Variant = Spanned<Variant_>;
 
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub enum UseTreeKind {
```
Comment (oli-obk) [inline src/libsyntax/ast.rs]: Can you add some docs?

**Review by oli-obk — APPROVED** [review]
r=me with some docs and a rebase

**Pull request merged**